controller/src/beerocks/master/wfa_ca.cpp

```diff
@@ -83,6 +123,46 @@ bool wfa_ca::reply(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx, const std::st
     return message_com::send_cmdu(sd, cmdu_tx);
 }
 
+bool wfa_ca::parse_params(std::unordered_map<std::string, std::string> &params,
+                          std::string &err_string)
+{
+    // create copy of all mandatory param names
+    std::list<std::string> mandatory_params;
+    for (const auto &param_name : params) {
+        mandatory_params.push_back(param_name.first);
+    }
+
+    size_t mandatory_params_cnt = 0;
+    while (!finished_tokens()) {
+        std::string param_name = get_next_token();
+        std::transform(param_name.begin(), param_name.end(), param_name.begin(), ::tolower);
+        if (params.find(param_name) != params.end()) {
+            mandatory_params_cnt++;
+        }
+
+        if (finished_tokens()) {
+            err_string = "missing param value for param '" + param_name + "'";
+            return false;
+        }
+
+        params[param_name] = get_next_token();
+    }
+
+    if (mandatory_params_cnt == mandatory_params.size()) {
+        return true;
+    }
+
+    // Generate err string if missing mandatory param
+    err_string = "missing mandatory params: ";
+    for (const std::string &param_name : mandatory_params) {
+        if ((params.find(param_name))->second.empty()) {
+            err_string += "'" + param_name + "', ";
+        }
+    }
+
```

Comment:
Since the caller always has the `LOG(ERROR)` and `reply_invalid` calls, it would make sense to those here instead.

Reply:
I'm sorry, can you please rephrase that - "it would make sense to those here instead"

Reply:
It would make sense to add those here. I.e., instead of sending the reply and doing the log in the caller, do it directly here.

However, I realise now that sending a reply from this function wouldn't be a good idea, because the other replies are sent from the caller. It is indeed not a good idea to sometimes send the reply from one function and sometimes from another. So it's fine as it is.